### PR TITLE
BUG: Index.equals raising with index of tuples that contain NA

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This came up when working on MultiIndex.union with ea. Gets converted into a Index of tuples and then raising the error.  The added check is a last resort before raising to avoid impacting performance negatively.

### CODE REVIEW

File: pandas/_libs/missing.pyx [hunk 1]
```diff
@@ -42,6 +42,34 @@ cdef:
...
+    -------
+    True if we are dealing with tuples that have NA on one side and non NA on
+    the other side.
+
+    """
+    if not isinstance(left, tuple) or not isinstance(right, tuple):
+        return False
+
+    for left_element, right_element in zip(left, right):
+        if left_element is C_NA and not right_element is C_NA:
```
Comment (jbrockmendel) [inline pandas/_libs/missing.pyx:65]: `not right_element is C_NA` -> `right_element is not C_NA`

File: pandas/_libs/lib.pyx [hunk 2]
```diff
@@ -659,7 +660,11 @@ def array_equivalent_object(left: object[:], right: object[:]) -> bool:
                   and not (isinstance(x, type(y)) or isinstance(y, type(x)))):
                 # Check if non-scalars have the same type
                 return False
+            elif check_na_tuples_nonequal(x, y):
+                # We have tuples where one Side has a NA and the other side does not
+                return False
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx:665]: comment that this is the only/main we we expect to get a TypeError

File: pandas/_libs/missing.pyx (hunk 1 above)
Comment (phofl) [reply pandas/_libs/missing.pyx]: thx

File: pandas/_libs/lib.pyx (hunk 2 above)
Comment (phofl) [reply pandas/_libs/lib.pyx:666]: Added

File: pandas/_libs/missing.pyx [hunk 3]
```diff
@@ -42,6 +42,34 @@ cdef:
...
+    Returns
+    -------
+    True if we are dealing with tuples that have NA on one side and non NA on
+    the other side.
+
+    """
+    if not isinstance(left, tuple) or not isinstance(right, tuple):
+        return False
+
+    for left_element, right_element in zip(left, right):
```
Comment (mroeschke) [inline pandas/_libs/missing.pyx:64]: Should this assert that `len(left) == len(right)`? Guessing this is assumed upstream but technically `(1, NA)` and `(1,)` would return `False`

Comment (phofl) [reply pandas/_libs/missing.pyx:67]: Added. I think we could end up here with weird combinations, so better to be a bit more explicit

**Review by mroeschke — APPROVED** [review]
LGTM

**Comment by phofl** [comment]
Ok to merge on greenish?

**Comment by mroeschke** [comment]
Thanks @phofl

File: doc/source/whatsnew/v1.6.0.rst [hunk 4]
```diff
@@ -168,7 +168,7 @@ Indexing
 
 Missing
 ^^^^^^^
--
+- Bug in :meth:`Index.equals` raising ``TypeError` when :class:`Index` consists of tuples that contain ``NA`` (:issue:`48446`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.6.0.rst:171]: looks like this is not rendering correctly due to unbalanced backticks around TypeError.

**Pull request merged**